File: frontend/src/services/taskService.ts
import apiClient from "@/lib/apiClient";
import type {
  TaskResponse,
  TaskUpdate,
  PageResponse,
  TaskParams,
} from "@/types";

export async function getTasks(
  params?: TaskParams
): Promise<PageResponse<TaskResponse>> {
  const { data } = await apiClient.get<PageResponse<TaskResponse>>(
    "/api/task",
    { params }
  );
  return data;
}

export async function getTaskById(id: number): Promise<TaskResponse> {
  const { data } = await apiClient.get<TaskResponse>(`/api/task/${id}`);
  return data;
}

export async function addTask(task: {
  taskName: string;
  taskDescription: string;
  taskStatus: string;
  dueDate: string;
  taskLabels: string[];
}): Promise<string> {
  const { data } = await apiClient.post<string>("/api/task", task);
  return data;
}

export async function updateTask(
  id: number,
  update: TaskUpdate
): Promise<string> {
  const { data } = await apiClient.put<string>(`/api/task/update/${id}`, update);
  return data;
}

export async function deleteTask(id: number): Promise<string> {
  const { data } = await apiClient.delete<string>(`/api/task/delete/${id}`);
  return data;
}

File: backend/src/main/java/com/justdeepfried/WorkStack/service/TaskService.java
package com.justdeepfried.WorkStack.service;

import com.justdeepfried.WorkStack.dto.PageResponse;
import com.justdeepfried.WorkStack.dto.Task.TaskResponse;
import com.justdeepfried.WorkStack.dto.Task.TaskUpdate;
import com.justdeepfried.WorkStack.exception.ResourceNotFound;
import com.justdeepfried.WorkStack.model.ActivityLogModel;
import com.justdeepfried.WorkStack.model.TaskModel;
import com.justdeepfried.WorkStack.repository.TaskRepo;
import com.justdeepfried.WorkStack.specification.TaskSpecification;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

@Service
public class TaskService {

    @Autowired
    private TaskRepo taskRepo;

    @Autowired
    private ActivityLogService activityLogService;

    public PageResponse<TaskResponse> getAll(Pageable pageable) {
        return PageResponse.from(taskRepo.findAll(pageable).map(TaskResponse::from));
    }

    public PageResponse<TaskResponse> getAll(Pageable pageable, String taskName) {
        Specification<TaskModel> spec = TaskSpecification.getSpecification(taskName);
        return PageResponse.from(taskRepo.findAll(spec, pageable).map(TaskResponse::from));
    }

    public TaskResponse getById(int taskId) {
        return TaskResponse.from(taskRepo.findById(taskId).orElseThrow(() -> new ResourceNotFound("Task with ID: " + taskId + " not found")));
    }

    public ResponseEntity<String> addTask(TaskModel taskModel) {
        taskRepo.save(taskModel);
        activityLogService.log(ActivityLogModel.Activity.CREATE, "Created a Task");
        return new ResponseEntity<>("Task added successfully!",HttpStatus.OK);
    }

    public ResponseEntity<String> deleteTask(int taskId) {
        TaskModel task = taskRepo.findById(taskId).orElseThrow(() -> new ResourceNotFound("Task with ID: " + taskId + " not found"));
        taskRepo.delete(task);
        activityLogService.log(ActivityLogModel.Activity.DELETE, "Deleted Task with ID: " + taskId);
        return new ResponseEntity<>("Task deleted successfully!",HttpStatus.OK);
    }

    public ResponseEntity<String> updateTask(int taskId, TaskUpdate taskUpdate) {
        TaskModel task = taskRepo.findById(taskId).orElseThrow(() -> new ResourceNotFound("Task with ID: " + taskId + " not found"));

        if (taskUpdate.taskName() != null) {
            task.setTaskName(taskUpdate.taskName());
        }
        if (taskUpdate.taskDescription() != null) {
            task.setTaskDescription(taskUpdate.taskDescription());
        }
        if (taskUpdate.taskStatus() != null) {
            task.setTaskStatus(taskUpdate.taskStatus());
        }
        if (taskUpdate.taskLabels() != null) {
            task.setTaskLabels(taskUpdate.taskLabels());
        }
        if (taskUpdate.dueDate() != null) {
            task.setDueDate(taskUpdate.dueDate());
        }

        taskRepo.save(task);
        activityLogService.log(ActivityLogModel.Activity.UPDATE, "Updated Task with ID: " + task.getTaskId());
        return new ResponseEntity<>("Task updated successfully!", HttpStatus.OK);
    }
}
